File: src/main/java/tomeko/chatblock/config/ChatBlockConfig.java
package tomeko.chatblock.config;

import dev.isxander.yacl3.api.*;
import dev.isxander.yacl3.api.controller.*;
import dev.isxander.yacl3.config.v2.api.ConfigClassHandler;
import dev.isxander.yacl3.config.v2.api.SerialEntry;
import dev.isxander.yacl3.config.v2.api.serializer.GsonConfigSerializerBuilder;
import dev.isxander.yacl3.platform.YACLPlatform;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.text.Text;
import tomeko.chatblock.utils.Constants;

import java.util.ArrayList;
import java.util.List;

public class ChatBlockConfig {
    public static final ConfigClassHandler<ChatBlockConfig> CONFIG = ConfigClassHandler.createBuilder(ChatBlockConfig.class)
            .serializer(config -> GsonConfigSerializerBuilder.create(config)
                    .setPath(YACLPlatform.getConfigDir().resolve(Constants.MOD_ID + ".json"))
                    .build())
            .build();

    @SerialEntry
    public static boolean blockSendingCaseSensitive = false;
    @SerialEntry
    public static List<String> messagesToBlockSending = new ArrayList<>();

    @SerialEntry
    public static boolean blockReceivingCaseSensitive = false;
    @SerialEntry
    public static boolean blockReceivingIgnoreFormatting = true;
    @SerialEntry
    public static List<String> messagesToBlockReceiving = new ArrayList<>();

    public static Screen configScreen(Screen parent) {
        return YetAnotherConfigLib.create(CONFIG, ((defaults, config, builder) -> builder
                .title(Text.literal(Constants.MOD_NAME))

                .category(ConfigCategory.createBuilder()
                        .name(Text.literal("Chat Block Config"))

                        .group(OptionGroup.createBuilder()
                                .name(Text.literal("Block Sending Custom Chat Messages"))
                                .option(Option.<Boolean>createBuilder()
                                        .name(Text.literal("Case-Sensitive"))
                                        .binding(defaults.blockSendingCaseSensitive, () -> config.blockSendingCaseSensitive, newVal -> config.blockSendingCaseSensitive = newVal)
                                        .controller(TickBoxControllerBuilder::create)
                                        .build())
                                .build())
                        .group(ListOption.<String>createBuilder()
                                .name(Text.literal("Block sending following chat messages:"))
                                .binding(defaults.messagesToBlockSending, () -> config.messagesToBlockSending, newVal -> config.messagesToBlockSending = newVal)
                                .controller(StringControllerBuilder::create)
                                .initial("")
                                .build())

                        .group(OptionGroup.createBuilder()
                                .name(Text.literal("Block Receiving Custom Chat Messages"))
                                .option(Option.<Boolean>createBuilder()
                                        .name(Text.literal("Case-Sensitive"))
                                        .binding(defaults.blockReceivingCaseSensitive, () -> config.blockReceivingCaseSensitive, newVal -> config.blockReceivingCaseSensitive = newVal)
                                        .controller(TickBoxControllerBuilder::create)
                                        .build())
                                .option(Option.<Boolean>createBuilder()
                                        .name(Text.literal("Ignore Formatting"))
                                        .binding(defaults.blockReceivingIgnoreFormatting, () -> config.blockReceivingIgnoreFormatting, newVal -> config.blockReceivingIgnoreFormatting = newVal)
                                        .controller(TickBoxControllerBuilder::create)
                                        .build())
                                .build())
                        .group(ListOption.<String>createBuilder()
                                .name(Text.literal("Block receiving following chat messages:"))
                                .binding(defaults.messagesToBlockReceiving, () -> config.messagesToBlockReceiving, newVal -> config.messagesToBlockReceiving = newVal)
                                .controller(StringControllerBuilder::create)
                                .initial("")
                                .build())
                        .build())

        )).generateScreen(parent);
    }

    public static void register() {
        ChatBlockConfig.CONFIG.load();
    }
}

File: src/main/kotlin/tomeko/chatblock/config/ChatBlockConfig.kt
package tomeko.chatblock.config

//? if = 1.8.9 {
/*import cc.polyfrost.oneconfig.config.Config
import cc.polyfrost.oneconfig.config.annotations.*
import cc.polyfrost.oneconfig.config.core.ConfigUtils
import cc.polyfrost.oneconfig.config.data.*
import cc.polyfrost.oneconfig.config.elements.*
import java.lang.reflect.Field
*///?} else {
import org.polyfrost.oneconfig.api.config.v1.Config
import org.polyfrost.oneconfig.api.config.v1.annotations.*
//?}
import tomeko.chatblock.utils.Constants

object ChatBlockConfig : Config(
    //? if = 1.8.9 {
    /*Mod(
        Constants.MOD_NAME,
        ModType.UTIL_QOL,
        "/assets/${Constants.MOD_ID}/icon.png"
    ),
    "${Constants.MOD_ID}.json"
    *///?} else {
    "${Constants.MOD_ID}.json",
    "/assets/${Constants.MOD_ID}/icon.png",
    Constants.MOD_NAME,
    Category.UTILITY
    //?}
) {
    fun register() {
        //? if = 1.8.9 {
        /*initialize()
        *///?} else {
        preload()
        clearPropertyLabels()
        //?}
    }

    //? if = 1.8.9 {
    /*@Exclude
    *///?}
    const val CATEGORY_BLOCK_RECEIVING = "Block Receiving Custom Messages"

    @Switch(
        //? if = 1.8.9 {
        /*name
            *///?} else {
            title
                //?}
        = "Case-sensitive",
        category = CATEGORY_BLOCK_RECEIVING
    )
    var blockReceivingCaseSensitive: Boolean = false

    @Switch(
        //? if = 1.8.9 {
        /*name
            *///?} else {
            title
                //?}
        = "Send message informing about a block",
        category = CATEGORY_BLOCK_RECEIVING
    )
    var blockReceivingInfoMessage: Boolean = false

    //? if >= 1.21.11 {
    @StringList(category = CATEGORY_BLOCK_RECEIVING)
    var messagesToBlockReceivingStringList: String = ""
    //?}

    //? if = 1.8.9 {
    /*@CustomOption(id = "blockReceiving")
    *///?}
    var messagesToBlockReceiving: Array<String> = emptyArray()
    //? if >= 1.21.11 {
    get() = messagesToBlockReceivingStringList.split("\n").filter { it.isNotBlank() }.toTypedArray()
//?}


    //? if = 1.8.9 {
    /*@Exclude
    *///?}
    const val CATEGORY_BLOCK_SENDING = "Block Sending Custom Words"

    @Slider(
        //? if = 1.8.9 {
        /*name
            *///?} else {
            title
                //?}
        = "Similarity",
        min = 1f,
        max = 100f,
        step =
            //? if = 1.8.9 {
            /*1
        *///?} else {
        1f
    //?}
        , category = CATEGORY_BLOCK_SENDING
    )
    var blockSendingSimilarity: Int = 100

    @Switch(
        //? if = 1.8.9 {
        /*name
            *///?} else {
            title
                //?}
        = "Send message informing about a block",
        category = CATEGORY_BLOCK_SENDING
    )
    var blockSendingInfoMessage: Boolean = true

    //? if >= 1.21.11 {
    @StringList(category = CATEGORY_BLOCK_SENDING)
    var wordsToBlockSendingStringList: String = ""
    //?}

    //? if = 1.8.9 {
    /*@CustomOption(id = "blockSending")
    *///?}
    var wordsToBlockSending: Array<String> = emptyArray()
    //? if >= 1.21.11 {
    get() = wordsToBlockSendingStringList.split(" ").filter { it.isNotBlank() }.toTypedArray()
    //?}


    //? if = 1.8.9 {
    /*@Exclude
    *///?}
    const val CATEGORY_DEBUG = "Debug"

    @Switch(
        //? if = 1.8.9 {
        /*name
            *///?} else {
            title
                //?}
        = "Debug Mode",
        category = CATEGORY_DEBUG
    )
    var debugModeEnabled = false

    //? if = 1.8.9 {
    /*override fun getCustomOption(
        field: Field,
        annotation: CustomOption,
        page: OptionPage,
        mod: Mod,
        migrate: Boolean
    ): BasicOption {
        when (annotation.id) {
            "blockReceiving" -> {
                val option = BlockReceivingListOption
                ConfigUtils.getSubCategory(page, CATEGORY_BLOCK_RECEIVING, "").options.add(option)
                return option
            }

            else -> {
                val option = BlockSendingListOption
                ConfigUtils.getSubCategory(page, CATEGORY_BLOCK_SENDING, "").options.add(option)
                return option
            }
        }
    }

    override fun load() {
        super.load()

        BlockReceivingListOption.apply {
            items.clear()
            items.addAll(messagesToBlockReceiving.map { message ->
                WrappedBlock(
                    message = message,
                    onRemove = { willBeRemoved = it },
                    splitOnSpace = false,
                    owner = BlockReceivingListOption
                )
            })
        }

        BlockSendingListOption.apply {
            items.clear()
            items.addAll(wordsToBlockSending.map { message ->
                WrappedBlock(
                    message = message,
                    onRemove = { willBeRemoved = it },
                    splitOnSpace = true,
                    owner = BlockSendingListOption
                )
            })
        }
    }

    override fun save() {
        messagesToBlockReceiving = BlockReceivingListOption.items
            .map { it.message }
            .toTypedArray()

        wordsToBlockSending = BlockSendingListOption.items
            .flatMap { it.message.split(Regex("\\s+")) }
            .filter { it.isNotBlank() }
            .toTypedArray()

        super.save()
    }
    *///?} else {
    private fun clearPropertyLabels() {
        try {
            var clazz: Class<*>? = this.javaClass
            while (clazz != null) {
                for (field in clazz.declaredFields) {
                    runCatching {
                        field.isAccessible = true
                        val value = field.get(this)
                        if (value != null) scanAndClean(value)
                    }
                }
                clazz = clazz.superclass
            }
        } catch (_: Throwable) {
        }
    }

    private fun scanAndClean(obj: Any) {
        val objClass = obj.javaClass
        val name = objClass.name

        if (obj is Map<*, *>) {
            obj.values.forEach { if (it != null) scanAndClean(it) }
            return
        }
        if (obj is Collection<*>) {
            obj.forEach { if (it != null) scanAndClean(it) }
            return
        }

        if (name.contains("polyfrost")) {
            runCatching {
                var c: Class<*>? = objClass
                var isTarget = false
                while (c != null) {
                    try {
                        val idField = c.getDeclaredField("id")
                        idField.isAccessible = true
                        val id = idField.get(obj)?.toString()
                        if (id == "messagesToBlockReceivingStringList" || id == "wordsToBlockSendingStringList") {
                            isTarget = true
                            break
                        }
                    } catch (_: Exception) {
                    }
                    c = c.superclass
                }

                if (isTarget) {
                    var currentClass: Class<*>? = objClass
                    while (currentClass != null) {
                        for (fieldName in listOf("title", "name", "label", "description")) {
                            try {
                                val f = currentClass.getDeclaredField(fieldName)
                                f.isAccessible = true
                                f.set(obj, "")
                            } catch (_: Exception) {
                            }
                        }
                        currentClass = currentClass.superclass
                    }
                }

                var scanClass: Class<*>? = objClass
                while (scanClass != null) {
                    for (field in scanClass.declaredFields) {
                        if (java.lang.reflect.Modifier.isStatic(field.modifiers)) continue
                        try {
                            field.isAccessible = true
                            val fieldVal = field.get(obj)
                            if (fieldVal != null && fieldVal !== obj && !field.type.isPrimitive) {
                                scanAndClean(fieldVal)
                            }
                        } catch (_: Exception) {
                        }
                    }
                    scanClass = scanClass.superclass
                }
            }
        }
    }
    //?}
}